### PR TITLE
Add Security Best Practices

### PR DESCRIPTION
**Pull request by robin-kunzler**
Adds security best practices for web app and canister development on the Internet Computer.

### CODE REVIEW

File: modules/security-best-practices/pages/web-app-development-security-best-practices.adoc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+=== Don’t use fetchRootKey in agent-js in production
+
+==== Security Concern
+
+`agent.fetchRootKey()` can be used in link:https://github.com/dfinity/agent-js[agent-js] to fetch the root subnet threshold public key from a status call in test environments. This key is used to verify threshold signatures on certified data received through canister update calls. Using this method in a production web app gives an attacker the option to supply their own public key, invalidating all authenticity guarantees of update responses.
+
+==== Recommendation
+
+Never use `agent.fetchRootKey()`only in test builds, but not in production builds. Not calling this method will result in the hard coded root subnet public key of mainnet being used for signature verification, which is the desired behavior in production.
```
Comment (sesi200) [inline modules/security-best-practices/pages/web-app-development-security-best-practices.adoc:36]:
```suggestion
Never use `agent.fetchRootKey()` in production builds, only in test builds. Not calling this method will result in the hard coded root subnet public key of mainnet being used for signature verification, which is the desired behavior in production.
```

Comment (robin-kunzler) [reply modules/security-best-practices/pages/web-app-development-security-best-practices.adoc]: Correct, thanks for spotting this @sesi200 ! Fixed it.

File: modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+=== Use a decentralized governance system like SNS to make a canister have a decentralized controller
+
+==== Security Concern
+
+The controller of a canister can change / update the canister whenever they like. If a canister e.g. stores assets such as ICP, this effectively means that the controller can steal these by updating the canister and transfer the cycles to their account.
+
+==== Recommendation
+
+To implement an immutable canister smart contract, you cannot be the controller of the canister. Control must be passed to the service nervous system (SNS) or a decentralized governance system.
```
Comment (dprats) [inline modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc:13]: is there not a third option? Blackhole the controller so the canister has no existing controller? would that not make it immutable? Also... this section should note that even immutable smart contracts NEED cycles (they are not permament unlike in other blockchains(, and that they can still receive cycles

File: modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+==== Recommendation
+
+- Consider end-to-end encrypting any private or personal data (e.g. user’s personal or private information) on canisters.
+- We are currently working on an example dApp (encrypted notes) to illustrate how end-to-end encryption can be done.
+
+=== Create backups
+
+==== Security Concern
+
+Bugs in the upgrade methods or inconsistent state could render a canister unusable.
```
Comment (dprats) [inline modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc:155]: Not sure it is quite clear whose bugs this is. Some people may think this refers to bugs in the IC, I would rewrite as: 

"If a canister has a faulty upgrade process (due to some bug from the dapp developer), then the canister can never be upgraded again. Make sure methods used in upgrading are tested or the canister becomes immutable."

**Review by dprats — COMMENTED** [review]
@robin-kunzler I only had two minor comments

File: modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc (hunk 3 above)
Comment (robin-kunzler) [reply modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc]: I agree this is unclear. Thanks for pointing this out. I changed it along the lines of your suggestion, hope it is clear now!

**Review by robin-kunzler — COMMENTED** [review]
@dprats : Thanks a lot for the review and the good feedback! I addressed your comments, see below.

**Pull request merged**